Improve the Getting Started on GOV.UK instructions (#2415)
alphagov/govuk-developer-docs: source/manual/get-started.html.md

    
1. Test that it all works by running `ssh -T git@github.com`.

1. While you're here, associate your name and email to your git commits:

    ```
    $ git config --global user.email "friendly.giraffe@digital.cabinet-office.gov.uk"
    $ git config --global user.name "Friendly Giraffe"
    ```

[GitHub]: https://www.github.com/
[user-reviewer]: https://github.com/alphagov/govuk-user-reviewer
[alphagov]: https://github.com/alphagov
[govuk-team]: https://github.com/orgs/alphagov/teams/gov-uk/members
[register-ssh-key]: https://help.github.com/articles/connecting-to-github-with-ssh/

## 2. Install GDS tooling

On GOV.UK we use two command-line tools day-to-day: [`govuk-connect`](https://github.com/alphagov/govuk-connect) and the [`gds-cli`](https://github.com/alphagov/gds-cli) for AWS, SSH and VPN access.

To install these, run:

```bash
brew tap alphagov/gds
brew install gds-cli govuk-connect
```

The GDS CLI repository is private, so you'll need to follow the GitHub setup instructions above for the download to work.

Test that both tools work by running `gds --help` and `gds govuk connect --help`.

> > If you see `fatal: no such path in the working tree`, it's because you're using ZSH, which has `gds` set up as a Git alias. You can either remove that alias by adding `unalias gds` to your `~/.zshrc`, or use `gds-cli` instead of `gds` for all the relevant commands.

Reviewer: I think that's probably on the case if you use oh-my-zsh or something - I use zsh and don't have (and haven't removed) a `gds` alias.

Author: Ah, maybe. But it's caught many people out in the workshops and was in the original docs. I say leave it? It's an "If", not a "Do this".

The GDS CLI requires some initial configuration:

```bash
gds config email firstname.lastname@digital.cabinet-office.gov.uk
gds config yubikey false # If you type MFA codes from your phone
```

## 3. Set up govuk-docker

We use a Docker environment - [govuk-docker][] - for local development.

👉 [Learn about how we use Docker](/manual/intro-to-docker.html)

> 👉 [Get setup with govuk-docker][govuk-docker]

Reviewer: Maybe link straight to https://github.com/alphagov/govuk-docker/blob/master/docs/installation.md?

```
class users::johnsmith {
  govuk_user { 'johnsmith':
    fullname => 'John Smith',
    email    => 'john.smith.123@digital.cabinet-office.gov.uk',
    ssh_key  => 'this public key will be a few lines long (copy the output from `more ~/.ssh/alphagov.pub`). It should begin with `ssh-rsa AAA` and end with `== john.smith.123@digital.cabinet-office.gov.uk`. You may need to add the email address to the end of your public key manually.',
  }
}
```

Add the name of your manifest (your username) into the list of `users::usernames` in [`hieradata_aws/integration.yaml`][integration-aws-hiera] for integration and in [`hieradata/integration.yaml`][integration-hiera] for CI.

Create a pull request with these changes. Once it has been [reviewed by a member of the GOV.UK team][merging], you can merge it and it will automatically deploy to the integration environment.

[govuk-puppet]: https://github.com/alphagov/govuk-puppet
[integration-aws-hiera]: https://github.com/alphagov/govuk-puppet/blob/master/hieradata_aws/integration.yaml
[integration-hiera]: https://github.com/alphagov/govuk-puppet/blob/master/hieradata/integration.yaml
[merging]: /manual/merge-pr.html

> ### Access remote environments

Reviewer: I think add somewhere in here that it can take up to 30 minutes after the PR has been merged

Your pull request from earlier will hopefully have been merged by now. It's time to test your access to servers via SSH.

> If you're not in the office right now, you'll need to be connected to the GDS office VPN for SSH access to integration.

Test that it works by running:

```bash
gds govuk connect --environment integration ssh backend
```

#### Running a console
Once you have SSH'd into a machine, you can also open a console for a particular application so you can execute commands, for example:

```bash

> govuk_app_console publishing-api

Reviewer: Make this one of the apps that live on a backend machine, like `signon`

Author: Oh, hah. Yes.

### Set up your MFA

You have to set up [Multi-Factor Authentication (MFA)][MFA].

1. [Sign in to AWS GDS account](https://gds-users.signin.aws.amazon.com/console)
2. Select or go to IAM service.
3. Click on "Users" in the menu bar on the left hand side
4. Enter your name
5. Click on the link for your email address
6. Click on the security credentials tab
7. Click on the "Manage" link next to "Assigned MFA device"
8. Follow the steps to set up your MFA device

> If you have a GDS-issued Yubikey, follow the [cross-GDS Yubikey docs](https://re-team-manual.cloudapps.digital/yubikeys.html#use-yubikey-for-2fa-in-amazon-web-services).

### Generate a pair of access keys

You have to generate an AWS Access Key and Secret Key to be able to
perform operations with AWS on the command-line.

1. [Sign in to the gds-users AWS Console](https://gds-users.signin.aws.amazon.com/console).
1. Click on your email address in the top right.
1. Click 'My Security Credentials'.
1. Click 'Create access key'.
1. Copy/paste them into the inputs that the gds-cli provides for you,
   if you're following [the setup instructions](#first-run).

#### Changing your MFA device

> 1. Follow steps 1 to 7 in [set up your MFA][#set-up-your-mfa]

Reviewer: Should be `[set up your MFA](#set-up-your-mfa)`

Author: Oops, thanks!

```
arn:aws:iam::<account-id>:user/<firstname.lastname>@digital.cabinet-office.gov.uk
```

After your PR has been merged, someone from the `govuk-administrators`
or `govuk-internal-administrators` group needs to deploy the
`infra-security` project.

## 6. Use your AWS access

### First run

Here, you will use the GDS CLI you installed and set up earlier. You'll be asked for AWS credentials on the first run:

```shell
$ gds aws govuk-integration-poweruser -l
Welcome to the GDS CLI! We will now store your AWS credentials in the keychain using aws-vault.
Enter Access Key ID: AK-YOUR-ACCESS-KEY-ID
Enter Secret Access Key: blah blah
Added credentials to profile "gds-users" in vault
Successfully initialised gds-cli
Enter token for arn:aws:iam::123456789012:mfa/firstname.lastname@digital.cabinet-office.gov.uk: 123456
```

Your (Secret) Access Key is from the AWS console. Follow [the instructions to generate one](#generate-a-pair-of-access-keys). The token requested at the end is the [MFA token](#set-up-your-mfa). If you have a GDS-issued Yubikey (you probably don't at this stage), set `gds config yubikey true` and the GDS CLI will automatically pull the MFA code from your Yubikey.

> ![AWS Access Key](images/aws/access-key.png)

Reviewer: This image doesn't exist

Author: Ah! It's in `source/images/aws/access-key.png` rather than `source/manual/images...`.

Author: Hmm, let's just delete this.

Test that both tools work by running `gds --help` and `gds govuk connect --help`.

> If you see `fatal: no such path in the working tree`, it's because you're using ZSH, which has `gds` set up as a Git alias. You can either remove that alias by adding `unalias gds` to your `~/.zshrc`, or use `gds-cli` instead of `gds` for all the relevant commands.

The GDS CLI requires some initial configuration:

```bash
gds config email firstname.lastname@digital.cabinet-office.gov.uk
gds config yubikey false # If you type MFA codes from your phone
```

## 4. Connecting to the GDS VPN

Access to our infrastructure and internal services is controlled by IP safelisting. If you're outside of the office or not on the Brattain network (ie, you're on GovWiFi), you'll need to connect to the GDS VPN to access our stuff.

To do this, read [the GDS Wiki page about the VPN][gds-vpn-wiki] to ensure you have the required pre-requisites, for example the Cisco AnyConnect software installed on your computer, and an MFA token given to you when you arrived by GDS IT. If you don't have these, [contact the IT helpdesk][gds-it-helpdesk].

The `gds-cli` tool that we installed earlier has a `vpn` subcommand.

> To connect to the VPN, run `gds vpn connect` and you'll be asked for your LDAP password and MFA code. To disconnect, run `gds vpn disconnect`.

Reviewer 2: What does LDAP stand for?

Author: LDAP is the protocol that we authenticate to GDS IT services with (ie when we login to our GDS laptops, print things, etc.) I wanted to say "laptop password", but for people on their own devices that's incorrect. I've asked IT what better words there are for this. Maybe "GDS IT password"? Will people understand that?

Author: ```suggestion
To connect to the VPN, run `gds vpn connect` and you'll be asked for a password and MFA code. The password is the one you use to login to your GDS-issued laptop or the printers.

To disconnect, run `gds vpn disconnect`.
```

Reviewer 2: I think GDS IT password is ok, I like your suggested change a lot though! Doesn't leave much room for doubt.

```
arn:aws:iam::<account-id>:user/<firstname.lastname>@digital.cabinet-office.gov.uk
```

After your PR has been merged, someone from the `govuk-administrators`
or `govuk-internal-administrators` group needs to deploy the
`infra-security` project.

## 8. Use your AWS access

### First run

Here, you will use the GDS CLI you installed and set up earlier. You'll be asked for AWS credentials on the first run:

```shell
$ gds aws govuk-integration-poweruser -l
Welcome to the GDS CLI! We will now store your AWS credentials in the keychain using aws-vault.
Enter Access Key ID: AK-YOUR-ACCESS-KEY-ID
Enter Secret Access Key: blah blah
Added credentials to profile "gds-users" in vault
Successfully initialised gds-cli
Enter token for arn:aws:iam::123456789012:mfa/firstname.lastname@digital.cabinet-office.gov.uk: 123456
```

Your (Secret) Access Key is from the AWS console. Follow [the instructions to generate one](#generate-a-pair-of-access-keys). The token requested at the end is the [MFA token](#set-up-your-mfa). If you have a GDS-issued Yubikey (you probably don't at this stage), set `gds config yubikey true` and the GDS CLI will automatically pull the MFA code from your Yubikey.

> ![AWS Access Key](images/aws/access-key.png)

Author: ```suggestion
```

[GDS]: https://gds.blog.gov.uk/about/
[overview-slides]: https://docs.google.com/presentation/d/1nAE65Og04JYNAc0VjYaUYLqNLuUOM9r3Mvo0PGFy_Zk

> ## 0. Install the [Homebrew package manager](https://brew.sh) (on macOS or Linux)

Reviewer 3: Zero-based indexing... I like it 😂 though should probably make this `1.` and update all the other numbers.

Author: Given I've basically updated every other number except the current `1.`, I suppose so.

## 1. Set up your GitHub account

> 1. Set up a [GitHub][] account. Existing personal accounts are fine to use.

Reviewer 3: I learned recently that you can omit the `[]`, so just `[GitHub]`.

1. Ask your tech lead to add you to the [alphagov organisation][alphagov]. You will have to be added to the [GOV.UK team][govuk-team] to get access to repos & CI. Remember to click accept in the GitHub email invitation.
1. Ask somebody with access to add your GitHub username to the [user monitoring system][user-reviewer].
1. [Generate and register an SSH key pair][register-ssh-key] for your laptop for your GitHub account. You should use a `4096` bit key.
1. Import the SSH key into your keychain. Once you’ve done this, you'll be able to clone repos over SSH.

    ```
    $ /usr/bin/ssh-add -K your-private-key
    ```

1. Add the above line into your `~/.bash_profile` or equivalent so that it is persistent between restarts.
    
1. Test that it all works by running `ssh -T git@github.com`.

1. While you're here, associate your name and email to your git commits:

    ```
    $ git config --global user.email "friendly.giraffe@digital.cabinet-office.gov.uk"
    $ git config --global user.name "Friendly Giraffe"
    ```

[GitHub]: https://www.github.com/
[user-reviewer]: https://github.com/alphagov/govuk-user-reviewer

> [alphagov]: https://github.com/alphagov

Reviewer 3: Could these be alphabetical?

[govuk-team]: https://github.com/orgs/alphagov/teams/gov-uk/members

> [register-ssh-key]: https://help.github.com/articles/connecting-to-github-with-ssh/

Author: ```suggestion
[alphagov]: https://github.com/alphagov
[GitHub]: https://www.github.com/
[govuk-team]: https://github.com/orgs/alphagov/teams/gov-uk/members
[register-ssh-key]: https://help.github.com/articles/connecting-to-github-with-ssh/
[user-reviewer]: https://github.com/alphagov/govuk-user-reviewer
```

👉 [Get setup with govuk-docker][govuk-docker]

[govuk-docker]: https://github.com/alphagov/govuk-docker

## 3. Get SSH access to integration

### Get access

Ask somebody with access to add your SSH username (`firstnamelastname`) to the [user monitoring system][user-reviewer].

[user-reviewer]: https://github.com/alphagov/govuk-user-reviewer

### Create a user to SSH into integration

User accounts in our integration environments are managed in the [govuk-puppet][] repository.

>     mac$ mkdir ~/govuk

Reviewer 3: We can remove the `mac$` prefixes now that we no longer need to distinguish between host commands and VM commands.

### Sign in to AWS

To sign in, go to [the GDS AWS Sign page](https://gds-users.signin.aws.amazon.com/console), and use the following credentials:

- "Account ID or alias": `gds-users`
- Username: your `@digital.cabinet-office.gov.uk` email address
- Password: your password

👉 [Sign in to AWS GDS account](https://gds-users.signin.aws.amazon.com/console)

### Set up your MFA

You have to set up [Multi-Factor Authentication (MFA)][MFA].

1. [Sign in to AWS GDS account](https://gds-users.signin.aws.amazon.com/console)
2. Select or go to IAM service.
3. Click on "Users" in the menu bar on the left hand side
4. Enter your name
5. Click on the link for your email address
6. Click on the security credentials tab
7. Click on the "Manage" link next to "Assigned MFA device"
8. Follow the steps to set up your MFA device

### Generate a pair of access keys

You have to generate an AWS Access Key and Secret Key to be able to
perform operations with AWS on the command-line.

> 1. [Sign in to the gds-users AWS Console](https://gds-users.signin.aws.amazon.com/console).

Reviewer 3: Can move this URL into an alias and call it from lines 164 and 178, rather than duplicate the URL - just makes it clearer to the doc maintainer that they refer to the same place.

Author: Sure!

Author: Done!

1. Click on your email address in the top right.

> 1. Click 'My Security Credentials'.

Reviewer 3: Having said that, these steps look identical up until this point 🤔 I wonder if they can be merged in some way?

Author: On this point, yes, but they're somewhat separate activities. I think it might be confusing to have to look elsewhere in this doc for how to even start doing the thing.

1. Click 'Create access key'.
1. Copy/paste them into the inputs that the gds-cli provides for you,
   if you're following [the setup instructions](/manual/access-aws-console.html).

#### Changing your MFA device

Follow steps 1 - 7 in [Set up your MFA](#3-set-up-your-mfa). Then:

1. Choose one of the two options (Remove or Resync)
2. Click on the "Manage" link next to "Assigned MFA device"

> 3. Follow the steps to set up your MFA device. If you're using a

Reviewer 3: These could change to `1.` prefixes.

👉 [Get setup with govuk-docker][govuk-docker]

> [govuk-docker]: https://github.com/alphagov/govuk-docker/blob/master/docs/installation.md

Reviewer 3: 👍
